dxw/dxw-members-only
# Feature/add http referrer allow list option

This commit adds the Referrer Allow list option, which can be used
to whitelist traffic arriving from certain referrers (internal only)
which is intended to allow plugins like Nelio AB Testing to be usable
on sites which use this plugin to control access to the site.

internal referrer urls can contain query strings, but the input
config should omit the site url, as this is set inside the plugin,
to prevent external referrers from being configured.

The new functionality also only accepts as valid referrer headers
where the configured allowed referrer string appears at the start
of the referrer header, to prevent whitelisted items being passed
as parameters of a referrer to circumvent the access controls.

loading an allow listed referrer url directly in the browser while
unauthenticated does not allow the user to bypass the access control
as when the plugin performs redirection it does not, itself, send the
HTTP_REFERER header.

- [x] Version number has been bumped

To test:
(the tests are unfortunately specific to judiciary intranet currently)
- in a local instance of judiciary-intranet, with nelio-ab-testing enabled and configured create a heatmap test for page http://localhost
- ensure that dxw-members only is configured to redirect visitors to the homepage to the login screen
- stop the test and attempt to view the results, this should time out with an error image (pictured below, first screenshot)
- in settings -> dxw members only add the url `/wp-admin/admin.php?page=nelio-ab-testing-experiment-view` to the Referrers Allow List config
- return to the nelio AB section and attempt to view the results of the test you created, the display should load as per screenshot 2 (below)
- log out of the local site and attempt to access the url `http://localhost/wp-admin/admin.php?page=nelio-ab-testing-experiment-view` you should be redirected back to the login screen.

Screenshot 1, non-allow-listed referrer behaviour
<img width="1342" alt="Screenshot 2023-10-23 at 10 24 33" src="https://github.com/dxw/dxw-members-only/assets/1659723/9eec20ce-ce9f-4900-8e72-9e42362a060e">

Screenshot 2, referrer allow listed behaviour
<img width="1400" alt="Screenshot 2023-10-23 at 10 27 39" src="https://github.com/dxw/dxw-members-only/assets/1659723/5029b66f-6300-4e0c-8b51-2a3655e3fb8b">
